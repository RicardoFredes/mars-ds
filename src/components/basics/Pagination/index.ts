/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 */

import Pagination from "./pagination.component";
export default Pagination;
export * from "./pagination.types";
