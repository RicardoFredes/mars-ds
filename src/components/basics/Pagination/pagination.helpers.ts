const DOTS = "...";
const FIRST_PAGE_INDEX = 1;

export const range = (start: number, end: number): number[] => {
  const length = end - start + 1;
  return Array.from({ length }, (_, idx) => idx + start);
};

export const getPaginationRange = (
  numPages: number,
  currentPage: number,
  siblingCount: number
): Array<number | string> => {
  const totalPageNumbers = siblingCount + 5;

  if (totalPageNumbers >= numPages) return range(1, numPages);

  const leftSiblingIndex = Math.max(currentPage - siblingCount, 1);
  const rightSiblingIndex = Math.min(currentPage + siblingCount, numPages);

  const shouldShowLeftDots = leftSiblingIndex > 2;
  const shouldShowRightDots = rightSiblingIndex < numPages - 2;

  const sideItemCount = 3 + 2 * siblingCount;

  if (!shouldShowLeftDots && shouldShowRightDots) return addRightDots(numPages, sideItemCount);

  if (shouldShowLeftDots && !shouldShowRightDots) return addLeftDots(numPages, sideItemCount);

  return addBothSideDots(numPages, leftSiblingIndex, rightSiblingIndex);
};

function addRightDots(numPages: number, sideItemCount: number) {
  const leftRange = range(1, sideItemCount);
  return [...leftRange, DOTS, numPages];
}

function addLeftDots(numPages: number, sideItemCount: number) {
  const rightRange = range(numPages - sideItemCount + 1, numPages);
  return [FIRST_PAGE_INDEX, DOTS, ...rightRange];
}

function addBothSideDots(numPages: number, leftSiblingIndex: number, rightSiblingIndex: number) {
  const middleRange = range(leftSiblingIndex, rightSiblingIndex);
  return [FIRST_PAGE_INDEX, DOTS, ...middleRange, DOTS, numPages];
}
