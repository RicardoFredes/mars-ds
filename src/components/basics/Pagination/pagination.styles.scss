.pagination {
  display: flex;
  list-style-type: none;
  padding: 0;
  justify-content: center;
  margin: 0;

  &__item {
    padding: 0 $spacing-md;
    height: 32px;
    text-align: center;
    margin: auto $spacing-xs;
    color: $color-text-primary;
    display: flex;
    box-sizing: border-box;
    align-items: center;
    border-radius: $border-radius-lg;
    font-size: $font-size-main;
    min-width: 32px;

    &--is-enabled {
      &:hover {
        background-color: $color-action-hover;
        cursor: pointer;
      }
    }

    &--is-current {
      background-color: $color-action-selected;
    }

    &--is-disabled {
      color: $color-action-disabled;
      pointer-events: none;
    }
  }
}
