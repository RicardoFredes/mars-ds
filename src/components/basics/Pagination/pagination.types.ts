export interface PaginationProps extends React.HTMLProps<HTMLUListElement> {
  defaultCurrentPage?: number;
  numPages?: number;
  siblingCount?: number;
}

export interface PaginationItemProps {
  pageNumber?: string | number;
  onClick?: VoidFunction;
  iconName?: string;
  isCurrent?: boolean;
  isDisabled?: boolean;
}
