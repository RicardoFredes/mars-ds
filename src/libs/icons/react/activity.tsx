/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconActivity = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12" />
</svg>
);
  
export default IconActivity;