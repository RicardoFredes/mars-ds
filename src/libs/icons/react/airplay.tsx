/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAirplay = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  stroke-linejoin="round"
>
  <path d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1" />
  <polygon points="12 15 17 21 7 21 12 15" />
</svg>
);
  
export default IconAirplay;