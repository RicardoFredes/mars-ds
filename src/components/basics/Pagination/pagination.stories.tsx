import type { ComponentMeta, ComponentStory } from "@storybook/react";

import { action } from "@storybook/addon-actions";

import Pagination from "./pagination.component";

export default {
  title: "Components/Pagination",
  component: Pagination,
  argTypes: {
    current: {
      description: "The current page",
      control: { type: "number" },
      defaultValue: 1,
    },
    total: {
      description: "The total number of pages",
      control: { type: "number" },
      defaultValue: 5,
    },
    siblingCount: {
      description: "The number of pages to show before and after the current page",
      control: { type: "number" },
      defaultValue: 2,
    },
    onSelectPage: {
      description: "The callback when page is selected",
      control: { type: "function" },
      action: action("onSelectPage"),
      defaultValue: undefined,
    },
  },
} as ComponentMeta<typeof Pagination>;

const Template: ComponentStory<typeof Pagination> = (args) => <Pagination {...args} />;

const DefaultArgs = {
  total: 5,
  siblingCount: 2,
};

export const Default = Template.bind({});
Default.args = DefaultArgs;
