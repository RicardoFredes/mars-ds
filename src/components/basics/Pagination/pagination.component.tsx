import type { PaginationProps, PaginationItemProps } from "./pagination.types";

import classNames from "classnames";
import { useState, useMemo } from "react";

import Icon from "@/components/basics/Icon";

import { getPaginationRange } from "./pagination.helpers";

const Pagination = ({
  className,
  defaultCurrentPage = 1,
  numPages = 5,
  siblingCount = 2,
  ...props
}: PaginationProps) => {
  const [currentPage, setCurrentPage] = useState(defaultCurrentPage);
  const paginationRange = useMemo(() => {
    return getPaginationRange(numPages, currentPage, siblingCount);
  }, [currentPage, numPages, siblingCount]);

  if (currentPage === 0 || paginationRange.length < 2) return null;

  const cn = classNames("pagination", className);

  const onPageChange = (page: number) => setCurrentPage(page);
  const onNext = () => onPageChange(currentPage + 1);
  const onPrevious = () => onPageChange(currentPage - 1);

  const lastPage = paginationRange[paginationRange.length - 1];
  return (
    <ul className={cn} {...props}>
      <PaginationItem
        data-testid="pagination-prev-button"
        iconName="chevron-left"
        isDisabled={currentPage === 1}
        onClick={onPrevious}
      />
      {paginationRange.map((pageNumber: number | string, index) => (
        <PaginationItem
          data-testid="pagination-page-number"
          key={index}
          isCurrent={pageNumber === currentPage}
          isDisabled={typeof pageNumber === "string"}
          onClick={() => onPageChange(Number(pageNumber))}
          pageNumber={pageNumber}
        />
      ))}
      <PaginationItem
        data-testid="pagination-next-button"
        iconName="chevron-right"
        isDisabled={currentPage === lastPage}
        onClick={onNext}
      />
    </ul>
  );
};

const PaginationItem = ({
  pageNumber,
  onClick,
  iconName,
  isCurrent,
  isDisabled,
  ...props
}: PaginationItemProps) => {
  const handleClick = () => {
    if (isDisabled) return;
    onClick?.();
  };

  return (
    <li
      className={classNames("pagination__item", {
        "pagination__item--is-current": isCurrent,
        "pagination__item--is-disabled": isDisabled,
        "pagination__item--is-enabled": onClick && !isDisabled,
      })}
      role="button"
      onClick={handleClick}
      {...props}
    >
      {iconName ? <Icon name={iconName} /> : pageNumber}
    </li>
  );
};

export default Pagination;
