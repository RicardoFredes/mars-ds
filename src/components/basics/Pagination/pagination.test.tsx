import type { PaginationProps } from "./pagination.types";

import { render } from "@testing-library/react";

import Pagination from "./pagination.component";

const dataTestId = {
  paginationPageNumber: "pagination-page-number",
  paginationPrevButton: "pagination-prev-button",
  paginationNextButton: "pagination-next-button",
};

const classNameLabel = {
  isCurrent: "pagination__item pagination__item--is-current pagination__item--is-enabled",
  isEnabled: "pagination__item pagination__item--is-enabled",
  isDisabled: "pagination__item pagination__item--is-disabled",
};

const makeSut = (props?: PaginationProps) => render(<Pagination {...props} />);

describe("<Pagination>", () => {
  it("should render component", () => {
    const wrapper = makeSut();
    expect(wrapper).toBeTruthy();
  });

  it("should render a default numbers of pages if not provided a numPages", () => {
    const wrapper = makeSut();
    const pageNumbers = wrapper.getAllByTestId(dataTestId.paginationPageNumber);
    expect(pageNumbers).toHaveLength(5);
  });

  it("should render the same amount of the numPages that are provided by numPages prop ", () => {
    const numPages = 3;
    const wrapper = makeSut({ numPages });
    const pages = wrapper.getAllByTestId(dataTestId.paginationPageNumber);
    expect(pages).toHaveLength(numPages);
  });

  it("should current page is equal to defaultCurrentPage prop", () => {
    const defaultCurrentPage = 5;
    const wrapper = makeSut({ defaultCurrentPage });
    const pages = wrapper.getAllByTestId(dataTestId.paginationPageNumber);

    expect(pages[0]?.className).toBe(classNameLabel.isEnabled);
    expect(pages[1]?.className).toBe(classNameLabel.isEnabled);
    expect(pages[2]?.className).toBe(classNameLabel.isEnabled);
    expect(pages[3]?.className).toBe(classNameLabel.isEnabled);
    expect(pages[4]?.className).toBe(classNameLabel.isCurrent);
  });

  it("should render a disabled previous button if the first element is selected", () => {
    const wrapper = makeSut();
    const pagePrevIcon = wrapper.getByTestId(dataTestId.paginationPrevButton);
    expect(pagePrevIcon.className).toBe(classNameLabel.isDisabled);
  });

  it("should render a enabled previous button if the first element is not selected", () => {
    const wrapper = makeSut({ defaultCurrentPage: 5 });
    const pagePrevIcon = wrapper.getByTestId(dataTestId.paginationPrevButton);
    expect(pagePrevIcon.className).toBe(classNameLabel.isEnabled);
  });

  it("should render a disabled next button if the last element is selected", () => {
    const wrapper = makeSut({ defaultCurrentPage: 5 });
    const pageNextIcon = wrapper.getByTestId(dataTestId.paginationNextButton);
    expect(pageNextIcon.className).toBe(classNameLabel.isDisabled);
  });

  it("should render a enabled next button if the last element is not selected", () => {
    const wrapper = makeSut();
    const pageNextIcon = wrapper.getByTestId(dataTestId.paginationNextButton);
    expect(pageNextIcon.className).toBe(classNameLabel.isEnabled);
  });
});
