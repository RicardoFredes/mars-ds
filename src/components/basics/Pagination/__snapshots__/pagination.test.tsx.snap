// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination> should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="pagination"
      >
        <li
          class="pagination__item pagination__item--is-disabled"
          data-testid="pagination-prev-button"
          role="button"
        >
          <span
            class="icon"
            data-name="chevron-left"
            data-testid="chevron-left"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
          </span>
        </li>
        <li
          class="pagination__item pagination__item--is-current pagination__item--is-enabled"
          data-testid="pagination-page-number"
          role="button"
        >
          1
        </li>
        <li
          class="pagination__item pagination__item--is-enabled"
          data-testid="pagination-page-number"
          role="button"
        >
          2
        </li>
        <li
          class="pagination__item pagination__item--is-enabled"
          data-testid="pagination-page-number"
          role="button"
        >
          3
        </li>
        <li
          class="pagination__item pagination__item--is-enabled"
          data-testid="pagination-page-number"
          role="button"
        >
          4
        </li>
        <li
          class="pagination__item pagination__item--is-enabled"
          data-testid="pagination-page-number"
          role="button"
        >
          5
        </li>
        <li
          class="pagination__item pagination__item--is-enabled"
          data-testid="pagination-next-button"
          role="button"
        >
          <span
            class="icon"
            data-name="chevron-right"
            data-testid="chevron-right"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
          </span>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="pagination"
    >
      <li
        class="pagination__item pagination__item--is-disabled"
        data-testid="pagination-prev-button"
        role="button"
      >
        <span
          class="icon"
          data-name="chevron-left"
          data-testid="chevron-left"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="15 18 9 12 15 6"
            />
          </svg>
        </span>
      </li>
      <li
        class="pagination__item pagination__item--is-current pagination__item--is-enabled"
        data-testid="pagination-page-number"
        role="button"
      >
        1
      </li>
      <li
        class="pagination__item pagination__item--is-enabled"
        data-testid="pagination-page-number"
        role="button"
      >
        2
      </li>
      <li
        class="pagination__item pagination__item--is-enabled"
        data-testid="pagination-page-number"
        role="button"
      >
        3
      </li>
      <li
        class="pagination__item pagination__item--is-enabled"
        data-testid="pagination-page-number"
        role="button"
      >
        4
      </li>
      <li
        class="pagination__item pagination__item--is-enabled"
        data-testid="pagination-page-number"
        role="button"
      >
        5
      </li>
      <li
        class="pagination__item pagination__item--is-enabled"
        data-testid="pagination-next-button"
        role="button"
      >
        <span
          class="icon"
          data-name="chevron-right"
          data-testid="chevron-right"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </span>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
